register vc duties with subnet tracker

arnetheduck opened the pull request:
* fix activation logging during startup
* cache slot signature to avoid duplicate signature work
* schedule aggregation duties one slot at a time to avoid CPU spike at
each epoch
* lower aggregation subnet pre-subscription time to 4 slots (lowers
bandwidth and CPU usage)

arnetheduck committed:
register vc duties with subnet tracker

* fix activation logging during startup
* cache slot signature to avoid duplicate signature work
* schedule aggregation duties one slot at a time to avoid CPU spike at
each epoch
* lower aggregation subnet pre-subscription time to 4 slots (lowers
bandwidth and CPU usage)

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -892,7 +659,18 @@ proc trackSyncCommitteeTopics*(node: BeaconNode) =
+  # these arrive
+  await node.registerDuties(slot)
+
   # Syncing tends to be ~1 block/s, and allow for an epoch of time for libp2p
```
Is this ~1 blocks/slot still true? It was an empirical measurement at the time.

arnetheduck replied:
eh, more like 4-5 blocks per second..

tersec commented on beacon_chain/validators/action_tracker.nim:
```diff
@@ -0,0 +1,198 @@
+  # Keep stability subnets for as long as validators are validating
+  var toPrune: seq[ValidatorIndex]
+  for k, v in tracker.knownValidators:
+    if v + SLOTS_PER_EPOCH * 3 < wallSlot: toPrune.add k
```
This is a little arbitrary, though reasonable -- how expensive are ENR updates?

arnetheduck replied:
the thinking here was that we in theory have a duty every epoch, but then we might not once we hit 260k, so we add an extra epoch just to be safe and another to avoid messy rounding

arnetheduck replied:
documented

tersec commented on beacon_chain/validators/action_tracker.nim:
```diff
@@ -0,0 +1,198 @@
+      ## Known aggregation duties in the near future - before each such
+      ## duty, we'll subscribe to the corresponding subnet to collect
+
+# https://github.com/ethereum/consensus-specs/blob/v1.0.1/specs/phase0/validator.md#phase-0-attestation-subnet-stability
```
There's still utility in noting where, yes, this functionality has existed in this form since v1.0.1, but the immediately relevant thing is that it's still this way, i.e. currently in v1.1.2.

arnetheduck replied:
fixed, copypaste thing

tersec commented on beacon_chain/nimbus_beacon_node.nim:
```diff
@@ -481,185 +485,27 @@ func toBitArray(stabilitySubnets: auto): BitArray[ATTESTATION_SUBNET_COUNT] =
-  for i in (prevAggregateSubnets - node.attestationSubnets.aggregateSubnets).
-      oneIndices():
-    node.attestationSubnets.subscribeSlot[i] = FAR_FUTURE_SLOT
+  if not node.actionTracker.subscribeAllSubnets:
```
`node.network.updateStabilitySubnetMetadata` effectively already checks this condition and will also early-exit

arnetheduck replied:
removed the if

arnetheduck committed:
fixes

* update stability subnets in ENR on startup
* log gossip state
* perform gossip subscriptions just before the next slot starts
* document stuff

arnetheduck committed:
add random include

arnetheduck committed:
fixes

* don't overwrite subscription state when not subscribed
* log target gossip state

arnetheduck committed:
fixes

* updating gossip status once is enough
* add test

arnetheduck committed:
fixes

* remove syncQueueLen - this one is not updated at the end of the sync
and may cause gossip to disconnect itself completely - use a simple head
distance instead
* fix gossip disconnection - if in hysteresis, node.gossipState will be
set to disabled even though we don't disable topic subscriptions
* fix extra duty registration call
